lib/logstash/string_interpolation.rb

```diff
@@ -0,0 +1,155 @@
+require "thread_safe"
+require "forwardable"
+
+module LogStash
+  class StringInterpolation
+    # Floats outside of these upper and lower bounds are forcibly converted
+    # to scientific notation by Float#to_s
+    MIN_FLOAT_BEFORE_SCI_NOT = 0.0001
+    MAX_FLOAT_BEFORE_SCI_NOT = 1000000000000000.0
+
+    INSTANCE = StringInterpolation.new
+    CACHE = ThreadSafe::Cache.new
+    TEMPLATE_TAG_REGEXP = /%\{[^}]+\}/
+
+    def evaluate(event, template)
+      if template.is_a?(Float) and
+          (template < MIN_FLOAT_BEFORE_SCI_NOT or template >= MAX_FLOAT_BEFORE_SCI_NOT) then
+        return template = ("%.15f" % template).sub(/0*$/,"")
+      end
+      
+      template = template.to_s
+
+      if not_cachable?(template)
+        return template
+      else
+        compiled = CACHE.get_or_default(template, nil) || CACHE.put(template, compile_template(template))
+        compiled.evaluate(event)
+      end
+    end
+
+    def self.evaluate(event, format)
+      INSTANCE.evaluate(event, format)
```

Comment:
in Ruby, using `extend self` on a module is the most elegant way to make a singleton class and you don't have to manage an `INSTANCE` variable. 

```ruby
module MySingleton
  extend self

  def public_function
    private_function
    puts(@instance_var)
  end

  private

  def private_function
    @instance_var ||= "foobar"
  end
end
```

There is always only one instance of `MySingleton` and you cannot call `new` on it thus you do not define the `initialize` method. You can use instance variables, but they must be initialized outside a constructor.